pallets/parachain-staking/src/lib.rs

```diff
@@ -0,0 +1,1163 @@
+			// Choose top TotalSelected collator candidates
+			let (v_count, collator_staked, delegator_staked) = <Pallet<T>>::select_top_candidates(0u32);
+			// Start Round 0 at Block 0
+			let round: RoundInfo<T::BlockNumber> = RoundInfo::new(0u32, 0u32.into(), T::DefaultBlocksPerRound::get());
+			<Round<T>>::put(round);
+			// Snapshot total stake
+			<Pallet<T>>::deposit_event(Event::NewRound(
+				T::BlockNumber::zero(),
+				0u32,
+				v_count,
+				collator_staked,
+				delegator_staked,
+			));
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Set the annual inflation rate to derive per-round inflation
+		#[pallet::weight(0)]
+		pub fn set_inflation(origin: OriginFor<T>, inflation: InflationInfo) -> DispatchResult {
+			frame_system::ensure_root(origin)?;
+
+			Self::update_inflation(inflation)?;
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		/// Set the total number of collator candidates selected per round
+		/// - changes are not applied until the start of the next round
+		pub fn set_total_selected(origin: OriginFor<T>, new: u32) -> DispatchResultWithPostInfo {
+			frame_system::ensure_root(origin)?;
+			ensure!(new >= T::MinSelectedCandidates::get(), Error::<T>::CannotSetBelowMin);
+			let old = <TotalSelected<T>>::get();
+			<TotalSelected<T>>::put(new);
+			Self::deposit_event(Event::TotalSelectedSet(old, new));
+			Ok(().into())
+		}
+
+		#[pallet::weight(0)]
+		/// Set blocks per round
+		/// - if called with `new` less than length of current round, will
+		///   transition immediately
+		/// in the next block
+		pub fn set_blocks_per_round(origin: OriginFor<T>, new: u32) -> DispatchResultWithPostInfo {
+			frame_system::ensure_root(origin)?;
+			ensure!(new >= T::MinBlocksPerRound::get(), Error::<T>::CannotSetBelowMin);
+
+			let mut round = <Round<T>>::get();
+			let (now, first, old) = (round.current, round.first, round.length);
+			round.length = new;
+			<Round<T>>::put(round);
+
+			Self::deposit_event(Event::BlocksPerRoundSet(now, first, old, new));
+			Ok(().into())
+		}
+
+		/// Join the set of collator candidates
+		#[pallet::weight(0)]
+		pub fn join_candidates(origin: OriginFor<T>, bond: BalanceOf<T>) -> DispatchResultWithPostInfo {
+			let acc = ensure_signed(origin)?;
+			ensure!(!Self::is_candidate(&acc), Error::<T>::CandidateExists);
+			ensure!(!Self::is_delegator(&acc), Error::<T>::DelegatorExists);
+			ensure!(bond >= T::MinCollatorCandidateStk::get(), Error::<T>::ValBondBelowMin);
+			ensure!(bond <= T::MaxCollatorCandidateStk::get(), Error::<T>::ValBondAboveMax);
+
+			let mut candidates = <CandidatePool<T>>::get();
+			ensure!(
+				candidates.insert(Bond {
+					owner: acc.clone(),
+					amount: bond
+				}),
+				Error::<T>::CandidateExists
+			);
+
+			// Post-launch TODO: Replace with `check_collator_candidate_inclusion`.
+			ensure!(
+				(candidates.len() as u32) <= T::MaxCollatorCandidates::get(),
+				Error::<T>::TooManyCollatorCandidates
+			);
+			T::Currency::reserve(&acc, bond)?;
```

Comment:
What happens if this call fails but the storage has been updated with the previous `candidates.insert()`? Should this also be transactional?

Reply:
Yeah I would like to have all extrinsics here be transactional. Unfortunately, I ran into compilation issues. If those cannot be resolved, we have to refactor the extrinsics to not make storage changes before potential errors!

Reply:
Tracking issue: https://github.com/kiltprotocol/ticket/issues/1335